util/topsql/reporter/pubsub.go

```diff
@@ -0,0 +1,267 @@
+
+package reporter
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/zap"
+)
+
+// TopSQLPubSubService implements tipb.TopSQLPubSubServer.
+//
+// If a client subscribes to TopSQL records, the TopSQLPubSubService is responsible
+// for registering an associated DataSink to the reporter. Then the DataSink sends
+// data to the client periodically.
+type TopSQLPubSubService struct {
+	dataSinkRegisterer DataSinkRegisterer
+}
+
+// NewTopSQLPubSubService creates a new TopSQLPubSubService.
+func NewTopSQLPubSubService(dataSinkRegisterer DataSinkRegisterer) *TopSQLPubSubService {
+	return &TopSQLPubSubService{dataSinkRegisterer: dataSinkRegisterer}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPubSubService{}
+
+// Subscribe registers dataSinks to the reporter and redirects data received from reporter
+// to subscribers associated with those dataSinks.
+func (ps *TopSQLPubSubService) Subscribe(_ *tipb.TopSQLSubRequest, stream tipb.TopSQLPubSub_SubscribeServer) error {
+	ds := newPubSubDataSink(stream, ps.dataSinkRegisterer)
+	if err := ps.dataSinkRegisterer.Register(ds); err != nil {
+		return err
+	}
+	return ds.run()
+}
+
+type pubSubDataSink struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	stream     tipb.TopSQLPubSub_SubscribeServer
+	sendTaskCh chan sendTask
+
+	// for deregister
+	registerer DataSinkRegisterer
+}
+
+func newPubSubDataSink(stream tipb.TopSQLPubSub_SubscribeServer, registerer DataSinkRegisterer) *pubSubDataSink {
+	ctx, cancel := context.WithCancel(stream.Context())
+
+	return &pubSubDataSink{
+		ctx:    ctx,
+		cancel: cancel,
+
+		stream:     stream,
+		sendTaskCh: make(chan sendTask, 1),
+
+		registerer: registerer,
+	}
+}
+
+var _ DataSink = &pubSubDataSink{}
+
+func (ds *pubSubDataSink) TrySend(data *ReportData, deadline time.Time) error {
+	select {
+	case ds.sendTaskCh <- sendTask{data: data, deadline: deadline}:
+		return nil
+	case <-ds.ctx.Done():
+		return ds.ctx.Err()
+	default:
+		ignoreReportChannelFullCounter.Inc()
+		return errors.New("the channel of pubsub dataSink is full")
+	}
+}
+
+func (ds *pubSubDataSink) OnReporterClosing() {
+	ds.cancel()
+}
+
+func (ds *pubSubDataSink) run() error {
+	defer func() {
+		ds.registerer.Deregister(ds)
+		ds.cancel()
+	}()
+
+	for {
+		select {
+		case task := <-ds.sendTaskCh:
+			ctx, cancel := context.WithDeadline(ds.ctx, task.deadline)
+			var err error
+
+			start := time.Now()
+			go util.WithRecovery(func() {
+				defer cancel()
+				err = ds.doSend(ctx, task.data)
+
+				if err != nil {
+					reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
+				} else {
+					reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+				}
+			}, nil)
+
+			// When the deadline is exceeded, the closure inside `go util.WithRecovery` above may not notice that
+			// immediately because it can be blocked by `stream.Send`.
```

Comment:
So if a stream is still blocked by existing sending operation and we are here continuing to process other send tasks, what happens next may be another `stream.Send()` being called (while the previous call has not returned). Will this cause problems?

Reply:
Never come to another `stream.Send()`.

3 cases trigger `<-ctx.Done()`:
* deadline is exceeded, then the `run` returns
* `stream.Send()` is done, then continue to wait for a next send task
* `ds.ctx.Done()`, then the `run` also returns.